sound/soc/sof/intel/hda-trace.c

```diff
@@ -67,27 +70,68 @@ int hda_dsp_trace_init(struct snd_sof_dev *sdev, u32 *stream_tag)
 	return ret;
 }
 
-int hda_dsp_trace_release(struct snd_sof_dev *sdev)
+static int hda_dma_trace_release(struct sof_client_dev *cdev)
 {
+	struct snd_sof_dev *sdev = sof_client_dev_to_sof_dev(cdev);
 	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
 	struct hdac_stream *hstream;
 
-	if (hda->dtrace_stream) {
-		hstream = &hda->dtrace_stream->hstream;
-		hda_dsp_stream_put(sdev,
-				   SNDRV_PCM_STREAM_CAPTURE,
-				   hstream->stream_tag);
-		hda->dtrace_stream = NULL;
-		return 0;
+	if (!hda->dtrace_stream) {
+		dev_dbg(sdev->dev, "DMA trace stream is not opened!\n");
+		return -ENODEV;
 	}
 
-	dev_dbg(sdev->dev, "DMA trace stream is not opened!\n");
-	return -ENODEV;
+	hstream = &hda->dtrace_stream->hstream;
+	hda_dsp_stream_put(sdev, SNDRV_PCM_STREAM_CAPTURE, hstream->stream_tag);
+	hda->dtrace_stream = NULL;
+
+	return 0;
 }
 
-int hda_dsp_trace_trigger(struct snd_sof_dev *sdev, int cmd)
+static int hda_dsp_trace_trigger(struct sof_client_dev *cdev, int cmd)
 {
+	struct snd_sof_dev *sdev = sof_client_dev_to_sof_dev(cdev);
 	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
 
 	return hda_dsp_stream_trigger(sdev, hda->dtrace_stream, cmd);
 }
+
+static int hda_dma_trace_start(struct sof_client_dev *cdev)
+{
+	return hda_dsp_trace_trigger(cdev, SNDRV_PCM_TRIGGER_START);
+}
+
+static int hda_dma_trace_stop(struct sof_client_dev *cdev)
+{
+	return hda_dsp_trace_trigger(cdev, SNDRV_PCM_TRIGGER_STOP);
+}
+
+static void hda_dma_trace_is_available(struct sof_client_dev *cdev,
+				       bool available)
+{
+	struct snd_sof_dev *sdev = sof_client_dev_to_sof_dev(cdev);
+
+	sdev->dtrace_is_available = available;
+}
+
+static const struct sof_dma_trace_host_ops hda_dma_trace_ops = {
+	.init = hda_dma_trace_init,
+	.release = hda_dma_trace_release,
+	.start = hda_dma_trace_start,
+	.stop = hda_dma_trace_stop,
+
+	.available = hda_dma_trace_is_available,
```

Comment:
this name sounds a bit like an "is this feature available?" test. Maybe `enable`?

Reply:
It is exactly for that purpose: to let the core know that we have loaded the module for the dtrace feature.
It is not to enable the dtrace, but for some reason the HDA needs to know that we have dtrace functionality enabled.
With dtrace in client, one can remove the module and insert it when it is needed and the core needs to know this.

It is more like that the dtrace support is available and not like enable or enabled.